sys/include/ztimer.h

```diff
@@ -0,0 +1,473 @@
+ *
+ * - the API is 32bit only
+ *
+ * - much of ztimer can be unittested on a mock clock
+ *
+ * - the internal timer list uses relative timers and thus doesn't need 64bit
+ *   arithmetic or storage
+ *
+ * - ztimer_extend extends <32bit timers in an ISR safe way
+ *
+ * - has much reduced number of configuration tunables (no XTIMER_BACKOFF,
+ *   XTIMER_ISR_BACKOFF)
+ *
+ * - ztimer always executes callbacks in ISR context, whereas xtimer might
+ *   trigger very short timers in ISR context
+ *
+ * @{
+ *
+ * @file
+ * @brief       ztimer API
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ * @author      Joakim Nohlgård <joakim.nohlgard@eistec.se>
+ */
+
+#ifndef ZTIMER_H
+#define ZTIMER_H
+
+#include <stdint.h>
+
+#include "kernel_types.h"
+#include "msg.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief ztimer_base_t forward declaration
+ */
+typedef struct ztimer_base ztimer_base_t;
+
+/**
+ * @brief ztimer_clock_t forward declaration
+ */
+typedef struct ztimer_clock ztimer_clock_t;
+
+/**
+ * @brief   Minimum information for each alarm
+ */
+struct ztimer_base {
+    ztimer_base_t *next;        /**< next alarm in list */
+    uint32_t offset;            /**< offset from last alarm in list */
+};
+
+/**
+ * @brief   ztimer structure
+ *
+ * This type represents an instance of an alarm, which is set on an
+ * underlying clock object
+ */
+typedef struct {
+    ztimer_base_t base;             /**< clock list entry */
+    void (*callback)(void *arg);    /**< alarm callback function pointer */
+    void *arg;                      /**< alarm callback argument */
+} ztimer_t;
+
+/**
+ * @brief   ztimer backend method structure
+ *
+ * This table contains pointers to the virtual methods for a ztimer clock.
+ *
+ * These functions used by ztimer core to interact with the underlying clock.
+ */
+typedef struct {
+    /**
+     * @brief   Set a new timer target
+     */
+    void (*set)(ztimer_clock_t *clock, uint32_t val);
+
+    /**
+     * @brief   Get the current count of the timer
+     */
+    uint32_t (*now)(ztimer_clock_t *clock);
+
+    /**
+     * @brief   Cancel any set target
+     */
+    void (*cancel)(ztimer_clock_t *clock);
+} ztimer_ops_t;
+
+/**
+ * @brief   ztimer device structure
+ */
+struct ztimer_clock {
+    ztimer_base_t list;             /**< list of active timers              */
+    const ztimer_ops_t *ops;        /**< pointer to methods structure       */
+    ztimer_base_t *last;            /**< last timer in queue, for _is_set() */
+    uint32_t adjust;                /**< will be subtracted on every set()  */
+#if MODULE_ZTIMER_EXTEND || DOXYGEN
+    /* values used for checkpointed intervals and 32bit extension */
+    uint32_t max_value;             /**< maximum relative timer value       */
+    volatile uint32_t checkpoint;   /**< cumulated time at last now() call  */
+    volatile uint32_t lower_last;   /**< timer value at last now() call     */
```

Comment:
Why `volatile` here?

Reply:
Thanks for pushing my nose in my usage of volatile. :wink: I'm unfortunately still in the habit of sticking volatile on every variable that's used in an ISR... I've removed the volatiles, as the memory barriers in isr_disable/restore should protect the variables (in `_ztimer_now_extend()`).